tensorly/decomposition/_parafac2.py

```diff
@@ -48,21 +48,26 @@ def _update_imputed(tensor_slices, mask, decomposition, method):
             slice_mean = tl.sum(slice * slice_mask) / (
                 tl.prod(slice.shape) - len(tl.where(slice_mask == 0)[0])
             )
+
             for i in range(slice.shape[0]):
                 for j in range(slice.shape[1]):
                     if slice_mask[i, j] == 0:
                         tl.index_update(
-                            tensor_slices, tl.index[slice_no, i, j], slice_mean
+                            tensor_slices[slice_no], tl.index[i, j], slice_mean
                         )
 
     else:  # factors
 
-        reconstructed_tensor = parafac2_to_tensor(decomposition)
+        reconstructed_slices = parafac2_to_slices(decomposition)
 
-        tensor_slices = (
-            mask * tensor_slices
-            + (tl.ones(shape=mask.shape) - mask) * reconstructed_tensor
-        )
+        for slice_no, (slice, rec_slice, slice_mask) in enumerate(
+            zip(tensor_slices, reconstructed_slices, mask)
+        ):
+
+            tensor_slices[slice_no] = (
+                slice_mask * slice
+                + (tl.ones(shape=slice_mask.shape) - slice_mask) * rec_slice
+            )
```

Comment:
If we end up using `tl.where` above, then we should probably also use it here for consistency :)

Reply:
Done! :)